src/mds/MDLog.cc

```diff
@@ -860,7 +861,6 @@ void MDLog::_trim_expired_segments()
     logger->inc(l_mdl_evtrm, ls->num_events);
     
     segments.erase(ls->seq);
-    delete ls;
```

Comment:
So my remaining concern with this PR is that it doesn't resolve the original ticket motivation:

> These really ought to be ref-counted in some way to **prevent early expiry.**

(Emphasis added.) We're still expiring the segment even if there may be outstanding references to it. The correct way to ensure there are no outstanding references would be to:

- lock a mutex associated with the map (from which clients obtain references), at this time that would be the mds_lock so no change really needed
- check the `LogSegmentRef` in the map has no oustanding references elsewhere in the MDS
- if so, then remove it from the map and delete

Reply:
Agree

Reply:
@batrick @tshacked Is this still true after you introduced the changes for checking SP use count when trimming expired segment?

Reply:
This looks resolved now!

Reply:
yes, there's no need to explicit delete the ls since it's being deleted by the SP